File: problem7/python-solution/p7.py
"""
p7.py

Zeke Reyna

Problem 7:
What is the 10001st prime number?
"""

import sys
import math

def generate_upper(n):
    term1 = n * math.log(n)
    term2 = n * math.log(math.log(n))
    return int(term1 + term2)

def generate_nums(nums, bound):
    for i in range(2, bound):
        nums.append(i)

def isFactor(num, fac):
    if num == fac:
        return False
    if num % fac == 0:
        return True
    return False # else

def sieve(nums, n):
    i = 0
    temp_n = 1
    while temp_n <= n and i < len(nums):
        temp_prime = nums[i]
        nums[:] = [x for x in nums if not isFactor(x, temp_prime)]
        i += 1
        temp_n += 1
    return nums[n-1]

if __name__ == "__main__":
    if len(sys.argv) == 2:
        n = int(sys.argv[1])
    else:
        n = 10001
    upperbound = generate_upper(n)
    nums = []
    generate_nums(nums, upperbound)
    nth_prime = sieve(nums, n)

    print ("\nThe {}th prime number is {}\n".format(n, nth_prime))

File: problem7/cpp-solution/p7.cpp
/*
 * p7.cpp
 *
 * Zeke Reyna
 *
 * Problem 7:
 * What is the 10001st prime number?
 */

#include <iostream>
#include <list>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;
const int _BOUND = 10001;

int generate_upper(int n)
{
	double term1, term2;
	term1 = n * log(n);
	term2 = n * log(log(n));
	return (int) (term1 + term2);
}

list<int> generate_nums(int bound)
{
	list<int> nums (bound - 2);
	iota(nums.begin(), nums.end(), 2);
	return nums;
}

int sieve(list<int>& nums, int n)
{
	int temp_prime = 1;
	int temp_n = 1;
	for (auto i1 = nums.begin(); i1 != nums.end() && temp_n <= n; ++i1) {
		temp_prime = *i1;
		for (auto i2 = next(i1); i2 != nums.end(); ) {
			if (*i2 % temp_prime == 0)
				i2 = nums.erase(i2);
			else
				++i2;
		}
		temp_n++;
	}
	return temp_prime;
}

int main(int argc, char **argv)
{
	ios_base::sync_with_stdio(false);
	int n;
	(argc == 2) ? (n = stoi(argv[1])) : (n = _BOUND);

	int upperbound = generate_upper(n);
	list<int> nums = generate_nums(upperbound);
	int nth_prime  = sieve(nums, n);

	cout << "\nThe " << n << "th prime number is " << nth_prime << "\n\n";

	return 0;
}
